Rework LayeredMaterial and Add Style

gchoqueux opened the pull request:
- **Add `Style` object to WFS**
- **Rework LayeredMaterial**
  - All textures and parameters are ordered by display order.
  - `this.Layers` contains all layers by display order
  - The slots's texture management is done internally, now, to set texture uses `layerId`
  - Fix bug : Remove color layer doesn't dispose textures colors
- **Uniform changes** (`LayeredMaterial` and `BasicMaterial`)
  - use`THREE.Uniform` in material
  - `int RTC`          becomes `bool useRTC`
  - `int lightingOn` becomes `bool lightingOn`
  - `int debug`       becomes `bool debug`
  - `int lightingOn` becomes `bool lightingOn`
  - `int selected`    becomes `bool selected`
- **Remove slot texture** in `NodeProcess` is replaced by `layerId`
- **Change Skirt size** influence performance rendering, minimize pixel shader pass
- **Fix typos and add Binary Tools**

gchoqueux committed:
Change Skirt size -> influence performance

peppsac commented on src/Core/Math/Binary.js:
```diff
@@ -0,0 +1,42 @@
+Binary.toChar = function(nMask) {
+  // nMask doit être compris entre -2147483648 et 2147483647
+  for (var nFlag = 0, nShifted = nMask, sMask = ""; nFlag < 32;
+       nFlag++, sMask += String(nShifted >>> 31), nShifted <<= 1);
```
Please rewrite this code, it's really obfuscated :-/
Also remove French comment.

peppsac commented on src/Core/Math/Binary.js:
```diff
@@ -0,0 +1,42 @@
+};
+
+Binary.ToInt32 = function(x) {
+        return x >> 0;
```
Javascript trick like this needs to be documented.
Please add a comment (from http://www.2ality.com/2012/02/js-integers.html):

```
Unsigned shifts convert their left-hand side to Uint32, signed shifts convert it to Int32.
Shifting by 0 digits returns the converted value. 
```

gchoqueux replied:
commented

peppsac commented on src/Core/Math/Binary.js:
```diff
@@ -0,0 +1,42 @@
+flag = flag | (offset);
+flag = flag | ((projection) << 5);
+flag = flag | ((visible) << 7 );
+flag = flag | (Binary.ToUint32(Math.floor(opacity * 255)) << 8);
```
What's the point of this `flag` variable? Also `flag` is a dull name - use something more descriptive.

gchoqueux replied:
It's example code

peppsac commented on src/Core/Math/Binary.js:
```diff
@@ -0,0 +1,42 @@
+// var g_offset= Binary.ToInt32(flag & Binary.ToUint32(32-1));
+// var g_projection = Binary.ToInt32((flag >> 5) & Binary.ToUint32(4-1));
+// var g_visible = Binary.ToInt32((flag >> 7) & Binary.ToUint32(2-1));
+// var g_opacity = Binary.ToInt32((flag >> 8) & Binary.ToUint32(256-1));
```
Deaaaaad code  :)

gchoqueux committed:
Add StyleManager object
Add Style object
Add Palette object

peppsac commented on src/Globe/TileGeometry.js:
```diff
@@ -129,6 +129,7 @@ TileGeometry.prototype.computeBuffers = function(params, builder) {
         };
     }
 
+    var id_m3,v1,v2,v3,v4;
```
I'd prefer to keep properly scoped variables using the `let` keyword

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -18,11 +18,13 @@ var emptyTexture = new THREE.Texture();
 var showDebug = false;
 var fooTexture;
 
+const ELEVATION = 0;
```
Nope. This is already defined in TileMesh (https://github.com/iTowns/itowns2/blob/704d4150cca62e918a8f7aeb99a393e52a02a778/src/Globe/TileMesh.js#L32)

Keep only one definition and `export` it so it can be used by other files as well.
And adapt the code everywhere to use them.
(doing so in a separate commit is probably a good idea)

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
+    // index = parseInt(index) || 0;
+    // index = index < 0 ? array.length + index : index;
+    // toIndex = parseInt(toIndex) || 0;
+    // toIndex = toIndex < 0 ? array.length + toIndex : toIndex;
```
Dead code.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
+    // index = index < 0 ? array.length + index : index;
+    // toIndex = parseInt(toIndex) || 0;
+    // toIndex = toIndex < 0 ? array.length + toIndex : toIndex;
+    toIndex = toIndex <= index ? toIndex : ((toIndex <= index + howMany) ? index + howMany : toIndex );
```
Please turn this in a multi-line statement to keep it readable.

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,63 +171,108 @@ LayeredMaterial.prototype.dispose = function() {
 
-LayeredMaterial.prototype.setSequence = function(newSequence) {
+        var layer = sequenceLayer[l];
+        var oldIndex = this.layers.indexOf(layer);
```
Replace `var` with `let` to get scope variables

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 
-LayeredMaterial.prototype.setLightingOn = function(enable) {
-    this.uniforms.lightingOn.value = enable === true ? 1 : 0;
+LayeredMaterial.prototype.setLayerFx = function(id, fx) {
```
Stop mixing id (as in layer.id = 'Ortho') and indices.
The 1st parameter of all the `setLayer*` must be `layerIndex`, or at least `idx` to make sure there's no confusion with `layer.id` (which is a string!)

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
-    return this.uniforms.nColorLayer.value - 1;
+LayeredMaterial.prototype.pushLayer = function(param) {
+
+    var l = this.layers.length;
```
`l` is a bad name. Use something more explicit like `layerIndex` or `index`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 
-LayeredMaterial.prototype.getNbColorTexturesLayer = function(layerIndex) {
-    return (this.paramLayers[layerIndex + 1].x || this.nbTextures[1]) - (this.paramLayers[layerIndex].x || 0);
+LayeredMaterial.prototype.getOffsetByIndex = function(index) {
```
`getOffsetByIndex` : `getTextureOffsetByLayerIndex` is longer  but explicit

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
-            param[l].tileMT === 'PM' ? 1 : 0, param[l].visible, param[l].opacity);
-        this.paramBLayers[l] = new THREE.Vector2(param[l].fx, 0.0);
-    }
+LayeredMaterial.prototype.getNbTexturesByIndex = function(index) {
```
`getNbTexturesByIndex` →`getTextureCountByLayerIndex`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
+    this.uniforms.visibility = new THREE.Uniform([true,true,true,true,true,true,true,true]);
+
+    // Array : Number total texture by type
+    this.uniforms.nbTextures = new THREE.Uniform(this.nbTextures);
```
`nbTextures` →`textureCount`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
-    this.paramLayers = Array(8);
-    this.paramBLayers = Array(8);
+    var paramLayers = Array(8);
+    this.nbTexturesLayers = Array(8);
```
`nbTexturesLayers` → `layersTextureCount`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
-            array[i] = remp;
-    };
+    // handle on textures number uniforms
+    this.nbTextures = [0,0];
```
`nbTextures` →`textureCount`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
+    // handle on textures uniforms
     this.Textures = [];
+    // handle on textures pitscale uniforms
     this.pitScale = [];
```
While cleanup variable names, `pitScale` could probably benefit from a better name. `offsetScale`?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
+};
+
+LayeredMaterial.prototype.setLayerOpacity = function(id, opacity) {
+    if(this.uniforms.paramLayers.value[id])
```
Why is this test done only in `setLayerOpacity`? Is it superfluous here or missing in other functions?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 };
 
 LayeredMaterial.prototype.setLayerVisibility = function(id, visible) {
+        this.uniforms.visibility.value[id] = visible;
```
Wrong indentation

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
-        this.paramLayers[id].z = visible ? 1 : 0;
-    }
+LayeredMaterial.prototype.setLayerNbTextures = function(id, nb) {
+        this.nbTexturesLayers[id] = nb;
```
Wrong indentation

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 
-LayeredMaterial.prototype.setNbLayersColor = function(n) {
-    this.uniforms.nColorLayer.value = n;
+LayeredMaterial.prototype.getNumberColorLayer = function() {
```
`getNumberColorLayer`→`getColorLayerCount`

peppsac commented on src/Globe/TileMesh.js:
```diff
@@ -228,30 +227,29 @@ TileMesh.prototype.setBBoxZ = function(min, max) {
 
-    return mat.getLevelLayerColor(l_COLOR, slot) < this.level;
-}
+    return mat.Textures[l_COLOR][slot].level < this.level;
```
Didn't you intend to hide the `slot` stuff inside LayeredMaterial?
Maybe the `downScaledLayer` function should live in LayeredMaterial?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -177,63 +171,108 @@ LayeredMaterial.prototype.dispose = function() {
+
+    // Dispose Layers textures
+    for (i = offset, max = offset + nbTextures; i < max; i++) {
+        if (this.Textures[COLOR][i] instanceof THREE.Texture)
```
Missing `{ }`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
-            this.paramLayers[l].z = 0;
-        }
+LayeredMaterial.prototype.setColorLayerParameters = function(params) {
+    for (var l = 0; l < params.length; l++) {
```
`setColorLayerParameters` is supposed to be used only once. So maybe add a few safety checks at the beginning:
- `loaded` should be false
-  `params.length` should be 0

gchoqueux replied:
some Params.length  can be equal to 0

peppsac replied:
My second line was wrong, I meant to write `this.getColorLayersCount() should be 0`

I still think you should add some safety checks.

gchoqueux replied:
Ok

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -41,124 +43,116 @@ var getColorAtIdUv = function(nbTex) {
     //this.wireframe = true;
 
-    this.layerIdToIndex = {};
+    this.layers = [];
```
Nope.
You're using it like this `this.layers.push(param.idLayer);` so `layersId` is a better name.
And since you're also doing `this.uniforms.nColorLayer.value = this.layers.length;` a better++ name is: `colorLayersId`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 };
 
-LayeredMaterial.prototype.getLevelLayerColor = function(id, slot) {
+LayeredMaterial.prototype.getLevelLayerColor = function(id, layer) {
```
This function was badely named. And so are the args.

WDYT of: `getColorLayerLevel = function(layerType, colorLayerId)`?

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -244,83 +283,87 @@ LayeredMaterial.prototype.setTexture = function(texture, layer, slot, pitScale)
 
-    var level = this.Textures[id][slot || 0].level;
+    var index = this.layers.indexOf(layer);
+    var slot = index > -1 ? this.getOffsetByIndex(index) : 0;
```
Maybe we should stop silently hiding errors.
`inded < 0` is a programming error so we should notify the programmer (`throw` maybe?)

peppsac commented on src/Scene/Description/Style.js:
```diff
@@ -0,0 +1,168 @@
+/**
+ * Generated On: 2015-10-5
+ * Class: Style
+ * Description: Description du style du Node. Le style et contenue sont les deux éléments permettant de convertir le Node pour le moteur graphique.
```
English only please.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -92,36 +92,43 @@ NodeProcess.prototype.subdivideNode = function(node, camera, params) {
 
                 child.matrixSet = [];
-
+                var tileMatrixSet,nTextures;
```
No. Use `let` instead of moving the variable out of their real scope.

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -92,36 +92,43 @@ NodeProcess.prototype.subdivideNode = function(node, camera, params) {
                         paramMaterial.push({
                             tileMT: tileMatrixSet,
-                            layerTexturesOffset: colorTextureCount,
+                            nTextures: nTextures,
```
`nTextures` could be replaced with `textureCount`

peppsac commented on src/Scene/NodeProcess.js:
```diff
@@ -292,7 +293,7 @@ function updateNodeImagery(quadtree, node, layersConfig, force) {
 
 function updateNodeElevation(quadtree, node, layersConfig, force) {
-    let currentElevation = node.materials[RendererConstant.FINAL].getLevelLayerColor(0, 0);
+    let currentElevation = node.materials[RendererConstant.FINAL].getLevelLayerColor(0);
```
If you use this function for elevation layer then you need to change it's name

peppsac commented on src/Scene/Scene.js:
```diff
@@ -87,14 +90,26 @@ Scene.prototype.getPickPosition = function(mouse) {
     return this.gfxEngine.getPickingPositionFromDepth(mouse);
 };
 
+Scene.prototype.addStyle = function(style) {
```
I'm not a fan of adding a ton of boiler-plate code, simply to forward method call.
Expose directly `stylesManager` instead.

peppsac reviewed (changes requested):
Corect me if I'm wrong, but the whole Pen/Style/Description is unused at the moment?
If yes then remove these change from this PR and re-submit them when you have a concrete usage example to submit.

gchoqueux committed:
Add binary tools object
Fix error parameter in WMTS
Remove warnings and fix typos

gchoqueux committed:
Rework LayeredMaterial and uniforms shaders
- Rework LayeredMaterial
  * All textures and parameters are ordered by display order.
  * `this.Layers` contains all layers by display order
  * The slots's texture management is done internally, now, to set texture uses `layerId`
  * Fix bug : Remove color layer doesn't dispose textures colors

- Uniform changes (`LayeredMaterial` and `BasicMaterial`)
  * use`THREE.Uniform` in material
  * `int RTC` becomes `bool useRTC`
  * `int lightingOn` becomes `bool lightingOn`
  * `int debug` becomes `bool debug`
  * `int lightingOn` becomes `bool lightingOn`
  * `int selected` becomes `bool selected`

peppsac commented on src/Renderer/LayeredMaterial.js:
```diff
@@ -16,23 +16,25 @@ import pitUV from 'Renderer/Shader/Chunk/pitUV.glsl';
 
 emptyTexture.level = -1;
-var nbLayer = 2;
+var layerTypesCount = 2;
```
Use `const` where possible

gchoqueux replied:
ok

gchoqueux commented:
Branch updated

gchoqueux committed:
fix problem

gchoqueux committed:
temp commit : resolve canceled command problem

gchoqueux commented:
Branch updated and rebased
